torch/linalg/__init__.py

```diff
@@ -13,3 +17,248 @@
+    pass
+
+@overload  # noqa: 749
+def norm(input, ord=None, dim=None, keepdim=False, out=None, dtype=None):  # noqa: 749
+    # type: (Tensor, str, Optional[int], bool, Optional[Tensor], Optional[int]) -> Tensor
+    pass
+
+def norm(input, ord=None, dim=None, keepdim=False, out=None, dtype=None):  # noqa: 749
+    r"""Returns the matrix norm or vector norm of a given tensor.
+
+    This function can calculate one of eight different types of matrix norms, or one
+    of an infinite number of vector norms, depending on both the number of reduction
+    dimensions and the value of the `ord` parameter.
+
+    Args:
+        input (Tensor): The input tensor. If dim is None, x must be 1-D or 2-D, unless :attr:`ord`
+            is None. If both :attr:`dim` and :attr:`ord` are None, the 2-norm of the input flattened to 1-D
+            will be returned.
+
+        ord (int, float, inf, -inf, 'fro', 'nuc', optional): The order of norm.
+            inf refers to :attr:`float('inf')`, numpy's :attr:`inf` object, or any equivalent object.
+            The following norms can be calculated:
+
+            =====  ============================  ==========================
+            ord    norm for matrices             norm for vectors 
+            =====  ============================  ==========================
+            None   Frobenius norm                2-norm
+            'fro'  Frobenius norm                -- not supported --
+            'nuc'  nuclear norm                  -- not supported --
+            inf    max(sum(abs(x), dim=1))       max(abs(x)) 
+            -inf   min(sum(abs(x), dim=1))       min(abs(x)) 
+            0      -- not supported --           sum(x != 0) 
+            1      max(sum(abs(x), dim=0))       as below 
+            -1     min(sum(abs(x), dim=0))       as below
+            2      2-norm (largest sing. value)  as below 
+            -2     smallest singular value       as below
+            other  -- not supported --           sum(abs(x)**ord)**(1./ord)
+            =====  ============================  ==========================
+
+            Default: ``None``
+
+        dim (int, 2-tuple of ints, 2-list of ints, optional): If :attr:`dim` is an int,
+            vector norm will be calculated over the specified dimension. If :attr:`dim`
+            is a 2-tuple of ints, matrix norm will be calculated over the specified
+            dimensions. If :attr:`dim` is None, matrix norm will be calculated
+            when the input tensor has two dimensions, and vector norm will be
+            calculated when the input tensor has one dimension. Default: ``None``
+
+        keepdim (bool, optional): If set to True, the reduced dimensions are retained
+            in the result as dimensions with size one. Default: ``False``
+
+        out (Tensor, optional): The output tensor. Ignored if ``None``. Default: ``None``
+
+        dtype (:class:`torch.dtype`, optional): The desired data type of
```

Comment:
This is a little misleading since, as this states in the next sentence, it's also the computation type.

Reply:
What happens if dtype and out are set and dtype != out.dtype?

Reply:
I see. If `dtype != out.dtype`, and error should be thrown. `torch.sum` has this behavior:
```
>>> a = torch.rand([10, 10]).double()
>>> b = torch.rand([10]).float()
>>> torch.sum(a, dim=0, dtype=torch.double, out=b)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: sum: provided dtype must match dtype of result. Got Float and Double.
```